### PR TITLE
[S2GRAPH-17]: Remove unnecessary abstraction layer, Storage.

### PR DESCRIPTION
**Pull request by SteamShon**
1. refactor all common codes from AsynchbaseStorage to Storage trait.
2. little bit of documentation.
3. restructuring package(move serializer/deserializer from storage/hbase into storage).
4. extract future cache from AsynchbaseStorage into utils.

### CODE REVIEW

File: s2rest_play/test/benchmark/JsonBenchmarkSpec.scala [hunk 1]
```diff
@@ -4,43 +4,43 @@ import play.api.libs.json.JsNumber
...
-    implicit val app = FakeApplication()
-
-    "json benchmark" in new WithApplication(app) {
+class JsonBenchmarkSpec extends BenchmarkCommon {
+  "to json" >> {
+    "json benchmark" >> {
 
       duration("map to json") {
-        (0 to 100) foreach { n =>
+        (0 to 10) foreach { n =>
```
Comment (SteamShon) [inline s2rest_play/test/benchmark/JsonBenchmarkSpec.scala]: I don`t think number of test here is not important so reduced them

File: s2rest_netty/src/main/scala/Server.scala [hunk 2]
```diff
@@ -130,11 +133,10 @@ class S2RestHandler(s2rest: RestHandler)(implicit ec: ExecutionContext) extends
         } else badRoute(ctx)
 
       case HttpMethod.POST =>
-        val jsonString = req.content.toString(CharsetUtil.UTF_8)
-        val jsQuery = Json.parse(jsonString)
+        val body = req.content.toString(CharsetUtil.UTF_8)
 
-        val result = s2rest.doPost(uri, jsQuery)
-        toResponse(ctx, req, jsQuery, result, startedAt)
+        val result = s2rest.doPost(uri, body, Option(req.headers().get(Experiment.impressionKey)))
```
Comment (SteamShon) [inline s2rest_netty/src/main/scala/Server.scala]: This is missed out part from S2GRAPH-45. so I merged them here.

File: loader/src/test/scala/subscriber/GraphSubscriberTest.scala [hunk 3]
```diff
@@ -28,7 +28,7 @@ class GraphSubscriberTest extends FunSuite with Matchers with WithKafka {
   test("GraphSubscriberHelper.store") {
     // actually we need to delete labelToReplace first for each test.
     val labelMapping = Map(testLabelName -> labelToReplace)
-    Management.copyLabel(testLabelName, labelToReplace, Some(hTableName))
+    GraphSubscriberHelper.management.copyLabel(testLabelName, labelToReplace, Some(hTableName))
```
Comment (SteamShon) [inline loader/src/test/scala/subscriber/GraphSubscriberTest.scala]: applied missing changes from Object to Class(Management).

File: loader/src/test/scala/subscriber/TransferToHFileTest.scala [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+import com.kakao.s2graph.core.Management
+import com.kakao.s2graph.core.types.HBaseType
+import org.apache.spark.{SparkConf, SparkContext}
+import org.scalatest._
+import subscriber.TransferToHFile._
+
+/**
+  * Created by Eric on 2015. 12. 2..
+  */
+class TransferToHFileTest  extends FlatSpec with BeforeAndAfterAll with Matchers {
```
Comment (SteamShon) [inline loader/src/test/scala/subscriber/TransferToHFileTest.scala]: Add Test cases for BulkLoader(TransferToHFile).

File: s2core/src/main/scala/com/kakao/s2graph/core/QueryParam.scala [hunk 5]
```diff
@@ -32,6 +32,11 @@ object Query {
   }
 }
 
+case class MultiQuery(queries: Seq[Query],
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/QueryParam.scala]: change location on file.

File: s2core/src/main/scala/com/kakao/s2graph/core/QueryResult.scala [hunk 6]
```diff
@@ -7,18 +7,22 @@ import scala.collection.Seq
...
-    val propsWithTs = Map(LabelMeta.timeStampSeq ->
-      InnerValLikeWithTs(InnerVal.withLong(currentTs, label.schemaVersion), currentTs))
-    for {
-      vertex <- query.vertices
-    } yield {
-      val edge = Edge(vertex, vertex, queryParam.labelWithDir, propsWithTs = propsWithTs)
-      val edgeWithScore = EdgeWithScore(edge, Graph.DefaultScore)
-      QueryRequestWithResult(QueryRequest(query, -1, vertex, queryParam),
-        QueryResult(edgeWithScoreLs = Seq(edgeWithScore)))
+    if (query.steps.isEmpty || query.steps.head.queryParams.isEmpty) {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/QueryResult.scala]: add short circuit to return empty result when query steps is empty.

File: s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala [hunk 7]
```diff
@@ -59,6 +57,7 @@ class RequestParser(config: Config) extends JSONParser {
 
   val hardLimit = 100000
   val defaultLimit = 100
+  val maxLimit = Int.MaxValue - 1
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala]: Integer overflow bug

File: s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala [hunk 8]
```diff
@@ -186,7 +183,8 @@ class RequestParser(config: Config) extends JSONParser {
       toQuery(queryJson, isEdgeQuery)
     }
     val weights = (jsValue \ "weights").asOpt[Seq[Double]].getOrElse(queries.map(_ => 1.0))
-    MultiQuery(queries = queries, weights = weights, queryOption = toQueryOption(jsValue))
+    MultiQuery(queries = queries, weights = weights,
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala]: change Query to accept original json value.

File: s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala [hunk 9]
```diff
@@ -432,31 +429,39 @@ class RequestParser(config: Config) extends JSONParser {
...
   }
 
-  def toEdge(jsValue: JsValue, operation: String) = {
 
-    val srcId = parse[JsValue](jsValue, "from") match {
-      case s: JsString => s.as[String]
-      case o@_ => s"${o}"
-    }
-    val tgtId = parse[JsValue](jsValue, "to") match {
+  private def toEdge(jsValue: JsValue, operation: String): List[Edge] = {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/rest/RequestParser.scala]: change toEdge return multiple edges.

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/IndexEdgeSerializable.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
+package com.kakao.s2graph.core.storage
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/IndexEdgeSerializable.scala]: move package from com.kakao.s2graph.core.storage.hbase to  com.kakao.s2graph.core.storage

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/MutationBuilder.scala [hunk 11]
```diff
@@ -1,57 +0,0 @@
-package com.kakao.s2graph.core.storage
-
-import com.kakao.s2graph.core._
-
-import scala.collection.Seq
-import scala.concurrent.ExecutionContext
-
-
-abstract class MutationBuilder[T](storage: Storage)(implicit ex: ExecutionContext) {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/MutationBuilder.scala]: remove unnecessary abstraction layer Mutation Builder.

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/QueryBuilder.scala [hunk 12]
```diff
@@ -1,104 +0,0 @@
...
-
-import com.google.common.cache.Cache
-import com.kakao.s2graph.core._
-import com.kakao.s2graph.core.types.{LabelWithDirection, VertexId}
-import com.kakao.s2graph.core.utils.logger
-import scala.collection.{Map, Seq}
-import scala.concurrent.{Future, ExecutionContext}
-import scala.util.Try
-
-abstract class QueryBuilder[R, T](storage: Storage)(implicit ec: ExecutionContext) {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/QueryBuilder.scala]: Also remove unnecessary abstract layer QueryBuilder.

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/SKeyValue.scala [hunk 13]
```diff
@@ -1,11 +1,28 @@
 package com.kakao.s2graph.core.storage
 
+import org.apache.hadoop.hbase.util.Bytes
 import org.hbase.async.KeyValue
 
-case class SKeyValue(table: Array[Byte], row: Array[Byte], cf: Array[Byte], qualifier: Array[Byte], value: Array[Byte], timestamp: Long) {
+
+object SKeyValue {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/SKeyValue.scala]: add Operation on SKeyValue.

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/Serializable.scala [hunk 14]
```diff
@@ -1,10 +1,10 @@
-package com.kakao.s2graph.core.storage.hbase
+package com.kakao.s2graph.core.storage
 
 import com.kakao.s2graph.core.storage.StorageSerializable
 
-object HSerializable {
+object Serializable {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/Serializable.scala]: Renaming

File: s2core/src/main/scala/com/kakao/s2graph/core/storage/Storage.scala [hunk 15]
```diff
@@ -1,64 +1,324 @@
...
+import org.apache.hadoop.hbase.util.Bytes
+import org.apache.kafka.clients.producer.ProducerRecord
+import scala.annotation.tailrec
 import scala.collection.Seq
+import scala.collection.mutable.ArrayBuffer
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.Try
+import scala.util.{Random, Try}
+
+abstract class Storage[R](val config: Config)(implicit ec: ExecutionContext) {
```
Comment (SteamShon) [inline s2core/src/main/scala/com/kakao/s2graph/core/storage/Storage.scala]: Major changes goes here. now all common codes located on Storage abstract class.

**Comment by daewon** [comment]
+1

**Comment by emeth-kim** [comment]
+1

**Comment by hsleep** [comment]
+1

**Pull request closed without merging**